### PR TITLE
ETD-394 Send messages to SQS queue for DSpace publication

### PR DESCRIPTION
**Pull request by jazairi**
#### Why these changes are being introduced:

In order to publish theses to DSpace, we need to send messages to
an SQS queue that the DSpace Submission Service (DSS) can read from
and continue the publication workflow.

#### Relevant ticket(s):

* https://mitlibraries.atlassian.net/browse/ETD-394

#### How this addresses that need:

This creates a DspacePublicationJob that constructs and sends
a DSS-compliant message to the SQS queue.

#### Side effects of this change:

* The AWS Ruby SDK requires message attribute keys to be snake case
instead of camel case and symbols instead of strings. It's unclear
if this will cause issues with DSS.
* Added a message to development.rb about the ActiveStorage host.
Currently this needs to be each time one starts a rails console,
but there is an open PR in the rails codebase that will allow us
to configure it for the dev environment.
* We chose to use AWS client stubs for testing purposes rather than
VCR. This means that we'll be instantiating the SQS client outside
of the job, but we don't anticipate any issues with this approach.

#### Developer

- [x] All new ENV is documented in README
- [ ] All new ENV has been added to Heroku Pipeline, Staging and Prod
- [x] ANDI or Wave has been run in accordance to
      [our guide](https://mitlibraries.github.io/guides/basics/a11y.html) and
      all issues introduced by these changes have been resolved or opened as new
      issues (link to those issues in the Pull Request details above)
- [ ] Stakeholder approval has been confirmed (or is not needed)

#### Code Reviewer

- [x] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The documentation has been updated or is unnecessary
- [ ] The changes have been verified
- [ ] New dependencies are appropriate or there were no changes

#### Requires database migrations?

NO

#### Includes new or updated dependencies?

YES

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -73,6 +73,15 @@ of variables:
 In addition, you will need to ensure the bucket CORS `AllowedOrigin` settings
 are configured to allow for the domain this app runs at.
 
+## SQS Configuration
+
+We use AWS SQS queues to publish theses to DSpace and read data about published theses from DSpace. The
+[DSpace Submission Service](https://github.com/MITLibraries/dspace-submission-service) middleware supports this workflow.
+
+`SQS_INPUT_QUEUE_NAME` - The name of the SQS input queue used for publication to DSpace.
```
[the remark this answers is not in the document]
Comment (JPrevost) [reply README.md:81]: The Python sdk allows us to use queues just by name. I want to see if the ruby sdk does as well as the full URL for the input queue and just the name of the output queue feels weird. It's fine if ruby doesn't have that affordance though.

File: config/environments/test.rb [hunk 2]
```diff
@@ -11,7 +11,9 @@
   ENV['SP_PRIVATE_KEY'] = ''
   ENV['SP_CERTIFICATE'] = ''
   ENV['THESIS_ADMIN_EMAIL'] = 'test@example.com'
-  ENV['OUTPUT_QUEUE_NAME'] = 'test_queue'
+  ENV['SQS_INPUT_QUEUE_URL'] = 'http://localhost:5000/123456789012/test1'
+  ENV['SQS_OUTPUT_QUEUE_NAME'] = 'test2'
```
[the remark this answers is not in the document]
Comment (jazairi) [reply config/environments/test.rb:15]: We need `SQS_OUTPUT_QUEUE_NAME` to build the message attributes. It's used in the SqsMessage class: https://github.com/MITLibraries/thing/blob/4f2b7bdf9c712f37eab8160c2db3d35a15c7d3ae/app/models/sqs_message.rb#L16

File: test/models/dspace_metadata_test.rb [hunk 3]
```diff
@@ -56,138 +70,251 @@ def dss_friendly_thesis(thesis)
...
+    u.orcid = nil
+    u.save
+    t2.reload
+    serialized = DspaceMetadata.new(t2).serialize_dss_metadata
+    unserialized = JSON.parse(serialized)
+
+    assert unserialized['metadata'].include?({ 'key' => 'dc.identifier.orcid', 'value' => '0001' })
+    refute unserialized['metadata'].include?({ 'key' => 'dc.identifier.orcid', 'value' => '0002' })
 
     # One author and no ORCID
```
Comment (matt-bernhardt) [inline test/models/dspace_metadata_test.rb:90]: I see what the changes to the test do, and don't object - but this is no longer a test of one author and no ORCIDs - there are two authors in the t2 fixture at this stage, neither of which have an ORCID. The tests seem to confirm that zero ORCIDs are present, which I think is the correct test, but the comment probably needs to be updated.

File: test/models/dspace_metadata_test.rb [hunk 4]
```diff
@@ -56,138 +70,251 @@ def dss_friendly_thesis(thesis)
...
-    assert_equal %w[MFA JD], dc['dc.description.degree']
-    assert_equal ['Master of Fine Arts', 'Master of Fine Arts'], dc['thesis.degree.name']
-    assert_equal %w[Bachelor Master], dc['mit.thesis.degree']
+    dss_friendly_thesis(t)
+    serialized = DspaceMetadata.new(t).serialize_dss_metadata
+    unserialized = JSON.parse(serialized)
+
+    assert unserialized['metadata'].include?({ 'key' => 'dc.description.degree', 'value' => 'MFA' })
+    assert unserialized['metadata'].include?({ 'key' => 'dc.description.degree', 'value' => 'JD' })
+    # TODO: are we supposed to be including the same degree name twice if they are the same?
```
Comment (matt-bernhardt) [inline test/models/dspace_metadata_test.rb:274]: Is there a ticket in the backlog to resolve this question?

File: test/models/dspace_metadata_test.rb [hunk 5]
```diff
@@ -10,43 +10,57 @@ def dss_friendly_thesis(thesis)
...
     t = theses(:one)
     dss_friendly_thesis(t)
-    dc = DspaceMetadata.new(t).instance_variable_get(:@dc)
-    assert_equal 'MyString', dc['dc.title']
-    assert_equal '2017-09', dc['dc.date.issued']
-    assert_equal 'MyText', dc['dc.description.abstract']
+    serialized = DspaceMetadata.new(t).serialize_dss_metadata
+    unserialized = JSON.parse(serialized)
+
+    assert_equal ['metadata'], unserialized.keys
```
Comment (matt-bernhardt) [inline test/models/dspace_metadata_test.rb:19]: This test is repeated down in the bottom of the file, and I'm not sure why - I'm not sure it is a huge problem, I've repeated tests as part of separate sequences to confirm the expected, but I'm not sure I follow the distinction between this test and what's on line 330.

**Review by matt-bernhardt — COMMENTED** [review]
I've looked at only the final commit in this PR, but no earlier work. If needed, I'm happy to look earlier if that would help give context. I'm making this a Comment rather than an Approval only because I want to make sure that I've done the expected scope of work.

The three comments I've made here are non-blocking in themselves. There's a comment in a test that I think needs to be updated, one repeated assertion which I'm not sure is needed, and a TODO comment that I think probably should become a ticket in the backlog.

File: test/models/dspace_metadata_test.rb (hunk 3 above)
Comment (JPrevost) [reply test/models/dspace_metadata_test.rb:90]: I suspect I copy/pastad the wrong comment. I'll take a closer look. Thanks for catching that.

**Pull request merged**